packages/docusaurus-plugin-client-redirects/src/writeRedirectFiles.ts

```diff
@@ -0,0 +1,86 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import fs from 'fs-extra';
+import path from 'path';
+import {memoize} from 'lodash';
+
+import {PluginContext, RedirectMetadata} from './types';
+import createRedirectPageContent from './createRedirectPageContent';
+import {
+  addTrailingSlash,
+  getFilePathForRoutePath,
+  removeTrailingSlash,
+} from './utils';
+
+export type WriteFilesPluginContext = Pick<PluginContext, 'baseUrl' | 'outDir'>;
+
+export type RedirectFileMetadata = {
+  fileAbsolutePath: string;
+  fileContent: string;
+};
+
+export function toRedirectFilesMetadata(
+  redirects: RedirectMetadata[],
+  pluginContext: WriteFilesPluginContext,
+): RedirectFileMetadata[] {
+  // Perf: avoid rendering the template twice with the exact same "props"
+  // We might create multiple redirect pages for the same destination url
+  // note: the first fn arg is the cache key!
+  const createPageContentMemoized = memoize((toUrl: string) => {
+    return createRedirectPageContent({toUrl});
+  });
+
+  const createFileMetadata = (redirect: RedirectMetadata) => {
+    const fileAbsolutePath = path.join(
+      pluginContext.outDir,
+      getFilePathForRoutePath(redirect.fromRoutePath),
+    );
+    const toUrl = addTrailingSlash(
+      `${removeTrailingSlash(pluginContext.baseUrl)}${path.join(
+        redirect.toRoutePath,
+      )}`,
+    );
```

Comment:
@yangshun @lex111 I'm not totally sure either if always having a trailing slash in redirect urls is good and supported by all host providers, I guess it's related to the FS redirect file location question as well.